### PR TITLE
[CMake] Add option to perform SwiftParser validation by default

### PR DESCRIPTION
**Pull request by rintaro**
Disabled by default. ~We'll enable it for some build presets.~

Enabled in `buildbot_swiftsyntax_{macos|linux}` and `source_compat_suite_{macos|linux}_*A` build presets

rdar://121545713

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: cmake/modules/AddPureSwift.cmake [hunk 1]
```diff
@@ -65,6 +65,12 @@ function(_add_host_swift_compile_options name)
   target_compile_options(${name} PRIVATE
     $<$<COMPILE_LANGUAGE:Swift>:-color-diagnostics>
   )
+
+  if(LLVM_ENABLE_ASSERTIONS)
+    target_compile_options(${name} PRIVATE "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-Xcc -UNDEBUG>")
+  else()
+    target_compile_options(${name} PRIVATE "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-Xcc -DNDEBUG>")
+  endif()
```
Comment (rintaro) [inline cmake/modules/AddPureSwift.cmake:73]: This is just a drive-by future proof fix. This ensures using the same `NDEBUG` between C/C++ and Swift.

**Comment by rintaro** [comment]
@swift-ci Please smoke test macOS

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by eeckstein — COMMENTED** [review]
lgtm, thanks!

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by bnbarham — APPROVED** [review]

**Pull request merged**